### ISSUE TITLE
Improving commands inserted with shortcuts

### ISSUE THREAD

**Issue by melsophos** [texstudio-org/texstudio#1013]
I am a bit puzzled by the default insertions of several commands using shortcuts:

1. The environment `equation` automatically inserts `\label{}` on the same line as `\begin{equation}`. First, I think one should not assume that every numbered equation should have a label and, second, I (personally) prefer when \label is on its own line.
2. No indentation is inserted in the line between the begin and end of the environment despite having the option "Indent and unindent automatically". I would say that an indentation should be automatically inserted when the indentation mode is "Keep indentation" or "Indent automatically".
3. I would have preferred to have a space inserted after the `%` sign (but I can imagine that other users prefer the contrary).
4. I could be great to have a shortcut which inserts `\left( \right)` as it is a quite common combination.

I can understand that providing too much control over the shortcuts would make the interface clumsy and the code heavier, but maybe default choices could be simpler. Then users can use macros for more complete commands.

**Comment by muzimuzhi** [comment]
> 1. The environment `equation` automatically inserts `\label{}` on the same line as `\begin{equation}`. First, I think one should not assume that every numbered equation should have a label and, second, I (personally) prefer when \label is on its own line.

I suppose you insert `equation` environment with menu item. You can change the corresponding inserted text under `Preferences -> (enable Advanced Options) -> Menus -> Math -> Math Equations -> env equation`,
![image](https://user-images.githubusercontent.com/6376638/77762521-9cf46100-7074-11ea-968f-ea951d9f0892.png)


> 2\. No indentation is inserted in the line between the begin and end of the environment despite having the option "Indent and unindent automatically". I would say that an indentation should be automatically inserted when the indentation mode is "Keep indentation" or "Indent automatically".

If you insert an environment by menu item `LaTeX -> Environments -> \begin{<environment>}`, then you can change the corresponding inserted text under similar Preference items.
![image](https://user-images.githubusercontent.com/6376638/77766519-b6001080-707a-11ea-8fdb-46176b90079c.png)


> 3\. I would have preferred to have a space inserted after the `%` sign (but I can imagine that other users prefer the contrary).

I want this feature too, but hesitated to post an issue. Maybe this can be achieved by defining a macro, like [spaceBeforeComment.txsMacro](https://github.com/texstudio-org/texstudio-macro/blob/master/automatedTextmanipulation/spaceBeforeComment.txsMacro).



> 4\. I could be great to have a shortcut which inserts `\left( \right)` as it is a quite common combination.

You can define a new menu item like
![image](https://user-images.githubusercontent.com/6376638/77766935-49394600-707b-11ea-8f82-f2ccd2a3409f.png)
And then allocate a shortcut for it in `Preferences -> Shortcuts`.

**Comment by melsophos** [comment]
Thanks a lot for your help!

Your comments solve completely my points 1) and 4). I was not aware of that menu, it's an amazing feature. I still think that \label should not be the default, but at least it's easy to solve.

> f you insert an environment by menu item `LaTeX -> Environments -> \begin{<environment>}`, then you can change the corresponding inserted text under similar Preference items.
> Note that leading spaces or tabs do affect the output in verbatim environments. Similar discussions: #815 .

Concerning point 2), your solution works very well (I just need to read the doc to find how to insert a tab instead of spaces), but only for this specific command. I find a bit cumbersome to edit all environments. I had indeed read #815 and found it related (and also a useful proposition by itself), but I still believe that indentation of environments should automatically follow the user preference of `indentation mode`.

> I want this feature too, but hesitated to post an issue. Maybe this can be achieved by defining a macro, like [spaceBeforeComment.txsMacro](https://github.com/texstudio-org/texstudio-macro/blob/master/automatedTextmanipulation/spaceBeforeComment.txsMacro).

I tried a naive simplification of the macro you point out (which is nice by the way): it is sufficient to write `(?language:latex)%` as a trigger, and `%% ` for the result. This works well when hitting the `%` key, but this does not work when inserting a comment with the shortcut. I guess that one should do something with the `editor:commentSelection()` call from the shortcut menu, I will have a look later. Another minor problem is that it cannot be used together `spaceBeforeComment.txsMacro`, I guess that one needs to design a single macro, checking if there is a text before or not, and then deciding what to insert (but this is not so important).
